rfd/0125-dynamic-auto-discovery-config.md

```diff
@@ -0,0 +1,406 @@
+  wrappers.LabelValues ResourceTags = 5;
+	// InstallParams sets the join method when installing on
+	// discovered Azure nodes.
+	MatcherInstallParams InstallParams = 6;
+}
+
+// GCPMatcher matches GCP resources.
+message GCPMatcher {
+	// Types are GKE resource types to match: "gke".
+	repeated string Types = 1;
+	// Locations are GKE locations to search resources for.
+	repeated string Locations = 2;
+	// Tags are GCP labels to match.
+  wrappers.LabelValues Tags = 3;
+	// ProjectIDs are the GCP project ID where the resources are deployed.
+	repeated string ProjectIDs = 4;
+}
+
+```
+
+### Backward Compatibility
+
+#### `discovery_service.selector`
+If the config has this field but Teleport version is no aware of that, the user must remove it before they can start the Discovery Service.
+
+#### Watcher on `DiscoveryConfig`
+If the `discovery_service` tries to set up a watcher over `DiscoveryConfig` but Teleport Auth/Proxy is not yet aware of this resource, then `discovery_service` will log a warning but continue with the static configuration.
+
+### UX
+
+#### `discovery_service` new service configuration properties
+Users need to add the `selector` property and then a list of label matchers.
+
+Example:
+```yaml
+discovery_service:
+  enabled: "yes"
+  selector:
+    labelA: value1 # matches against a single value
+    labelB: [value2, value3] # matches against multiple values
+    labelC: "*" # matches on all values
+```
+
+#### WebAPI: manage `DiscoveryConfig` resource
+The following endpoints must be create to be used by WebAPI:
+```
+Methods:
+GET .../discoveryconfig
+GET .../discoveryconfig/:id
+POST .../discoveryconfig
+PUT .../discoveryconfig/:id
+DELETE .../discoveryconfig/:id
+```
+
+JSON representation:
+```json
+{
+   "aws":[
+      {
+         "types":[
+            "ec2"
+         ],
+         "regions":[
+            "us-east-1",
+            "us-west-1"
+         ],
+         "tags":{
+            "*":"*"
+         },
+         "install":{
+            "join_params":{
+               "token_name":"aws-discovery-iam-token"
+            },
+            "script_name":"default-installer"
+         },
+         "ssm":{
+            "document_name":"TeleportDiscoveryInstaller"
+         }
+      }
+   ],
+   "azure":[
+      {
+         "types":[
+            "aks"
+         ],
+         "regions":[
+            "eastus",
+            "westus"
+         ],
+         "subscriptions":[
+            "11111111-2222-3333-4444-555555555555"
+         ],
+         "resource_groups":[
+            "group1",
+            "group2"
+         ],
+         "tags":{
+            "*":"*"
+         }
+      }
+   ],
+   "gcp":[
+      {
+         "types":[
+            "gke"
+         ],
+         "locations":[
+            "*"
+         ],
+         "tags":{
+            "*":"*"
+         },
+         "project_ids":[
+            "myproject"
+         ]
+      }
+   ]
+}
+```
+
+#### `tctl`: manage `DiscoveryConfig` resource
```

Comment:
Do we plan to support  discovery service/servers agents listing ?

Reply:
I think we should be able to list every agent/service :+1: 

I was not considering adding those listings as part of this RFD